x/evm/keeper/state_transition.go

```diff
@@ -324,31 +323,12 @@ func (k *Keeper) ApplyMessageWithConfig(ctx sdk.Context,
 	stateDB := statedb.New(ctx, k, txConfig)
 	evm := k.NewEVM(ctx, msg, cfg, tracer, stateDB)
 
-	// set the custom precompiles to the EVM (if any)
-	if cfg.Params.HasCustomPrecompiles() {
-		customPrecompiles := cfg.Params.GetActivePrecompilesAddrs()
-
-		activePrecompiles := make([]common.Address, len(vm.PrecompiledAddressesBerlin)+len(customPrecompiles))
-		copy(activePrecompiles[:len(vm.PrecompiledAddressesBerlin)], vm.PrecompiledAddressesBerlin)
-		copy(activePrecompiles[len(vm.PrecompiledAddressesBerlin):], customPrecompiles)
-
-		// Check if the transaction is sent to an inactive precompile
-		//
-		// NOTE: This has to be checked here instead of in the actual evm.Call method
-		// because evm.WithPrecompiles only populates the EVM with the active precompiles,
-		// so there's no telling if the To address is an inactive precompile further down the call stack.
-		toAddr := msg.To()
-		if toAddr != nil &&
-			slices.Contains(types.AvailableEVMExtensions, toAddr.String()) &&
-			!slices.Contains(activePrecompiles, *toAddr) {
-			return nil, errorsmod.Wrap(types.ErrInactivePrecompile, "failed to call precompile")
-		}
-
-		// NOTE: this only adds active precompiles to the EVM.
-		// This means that evm.Precompile(addr) will return false for inactive precompiles
-		// even though this is actually a reserved address.
-		precompileMap := k.Precompiles(activePrecompiles...)
-		evm.WithPrecompiles(precompileMap, activePrecompiles)
+	// Set the custom precompiles to the EVM if:
+	// 1. there are custom precompiles
+	// 2. the message is a contract call
+	if cfg.Params.HasCustomPrecompiles() && types.IsContractCall(msg) {
+		activePrecompiles, activePrecompilesMap := k.GetActivePrecompilesInstances(ctx, cfg.Params)
+		evm.WithPrecompiles(activePrecompilesMap, activePrecompiles)
```

Comment:
why can't we check the `to` address instead of instantiating all precompiles on every state transition

Reply:
what if it's a smart contract that's calling multiple ERC-20 extensions?

Reply:
It will be very costly to also track all subsequent contract calls given the max depth is set to 1024.